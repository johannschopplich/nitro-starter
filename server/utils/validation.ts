import type { H3Event } from 'h3'
import * as v from 'valibot'

const DEFAULT_ERROR_STATUS = 400
const DEFAULT_ERROR_MESSAGE = 'Validation Error'

function createValidationError(error: any) {
  return createError({
    statusCode: DEFAULT_ERROR_STATUS,
    statusMessage: DEFAULT_ERROR_MESSAGE,
    message: error?.message || DEFAULT_ERROR_MESSAGE,
    data: error,
  })
}

export function useValidatedQuery<
  T extends v.BaseSchema<unknown, unknown, v.BaseIssue<unknown>>,
>(event: H3Event, schema: T): v.InferOutput<T> {
  try {
    const query = getQuery(event)
    const parsed = v.parse<T>(schema, query)
    return parsed
  } catch (error) {
    console.error(error)
    throw createValidationError(error)
  }
}

export async function useValidatedBody<
  T extends v.BaseSchema<unknown, unknown, v.BaseIssue<unknown>>,
>(event: H3Event, schema: T): Promise<v.InferOutput<T>> {
  try {
    const body = await readBody(event)
    const parsed = v.parse<T>(schema, body)
    return parsed
  } catch (error) {
    console.error(error)
    throw createValidationError(error)
  }
}
