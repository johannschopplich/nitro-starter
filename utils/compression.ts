import { promisify } from 'node:util'
import zlib from 'node:zlib'
import type { Buffer } from 'node:buffer'
import type { H3Event } from 'h3'

export async function maybeCompressedSend(event: H3Event, response: Buffer) {
  const encodingMethod = getAnyCompression(event)

  if (encodingMethod) {
    const compression = promisify(
      zlib[encodingMethod === 'br' ? 'brotliCompress' : encodingMethod],
    )
    setResponseHeader(event, 'Content-Encoding', encodingMethod)
    return send(event, await compression(response))
  }

  return send(event, response)
}

function getAnyCompression(event: H3Event) {
  const encoding = getRequestHeader(event, 'accept-encoding')

  if (encoding?.includes('br')) return 'br'
  if (encoding?.includes('gzip')) return 'gzip'
  if (encoding?.includes('deflate')) return 'deflate'

  return undefined
}
