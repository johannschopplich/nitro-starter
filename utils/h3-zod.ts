// Adapted from https://github.com/wobsoriano/h3-zod

import { createError, eventHandler, getQuery, isMethod, readBody } from 'h3'
import type { EventHandler, H3Event } from 'h3'
import type { z } from 'zod'

// Duplicate private Zod utility type
type UnknownKeysParam = 'passthrough' | 'strict' | 'strip'

type Refined<T extends z.ZodType> = T extends z.ZodType<infer O>
  ? z.ZodEffects<T, O, O>
  : never

export type IOSchema<U extends UnknownKeysParam = any> =
  | z.ZodObject<any, U>
  | z.ZodUnion<[IOSchema<U>, ...IOSchema<U>[]]>
  | z.ZodIntersection<IOSchema<U>, IOSchema<U>>
  | z.ZodDiscriminatedUnion<string, z.Primitive, z.ZodObject<any, U>>
  | Refined<z.ZodObject<any, U>>

export function useValidatedQuery<T extends IOSchema>(
  event: H3Event,
  schema: T,
  /** Use your own error handler instead of throwing an H3Error */
  errorHandler?: (error: z.ZodIssue[]) => void,
) {
  const query = getQuery(event)
  const parsed = schema.safeParse(query)

  if (!parsed.success) {
    if (errorHandler) {
      errorHandler(parsed.error.issues)
      return
    }

    throw createError({
      statusCode: 400,
      statusMessage: JSON.stringify({
        errors: parsed.error.issues,
      }),
    })
  }

  return parsed.data as z.infer<T>
}

export async function useValidatedBody<T extends IOSchema>(
  event: H3Event,
  schema: T,
  /** Use your own error handler instead of throwing an H3Error */
  errorHandler?: (error: z.ZodIssue[]) => void,
) {
  const body = await readBody(event)
  const parsed = schema.safeParse(body)

  if (!parsed.success) {
    if (errorHandler) {
      errorHandler(parsed.error.issues)
      return
    }

    throw createError({
      statusCode: 400,
      statusMessage: JSON.stringify({
        errors: parsed.error.issues,
      }),
    })
  }

  return parsed.data as z.infer<T>
}

interface RequestSchemas<TBody extends IOSchema, TQuery extends IOSchema> {
  body?: TBody
  query?: TQuery
}

interface RequestIssues {
  body: z.ZodIssue[] | null
  query: z.ZodIssue[] | null
}

export function defineEventHandlerWithSchema<
  TBody extends IOSchema,
  TQuery extends IOSchema,
>({
  handler,
  schema,
  errorHandler,
}: {
  handler: EventHandler
  schema: RequestSchemas<TBody, TQuery>
  errorHandler?: (errors: RequestIssues, event: H3Event) => void
}) {
  return eventHandler(async (event) => {
    const errors: RequestIssues = {
      body: null,
      query: null,
    }

    const parsedData: {
      body: z.infer<TBody> | null
      query: z.infer<TQuery> | null
    } = {
      body: null,
      query: null,
    }

    if (schema.query) {
      const query = getQuery(event)
      const parsed = schema.query.safeParse(query)

      if (!parsed.success) errors.query = parsed.error.issues
      else parsedData.query = parsed.data as z.infer<TQuery>
    }

    if (schema.body && isMethod(event, 'POST')) {
      const body = await readBody(event)
      const parsed = schema.body.safeParse(body)

      if (!parsed.success) errors.body = parsed.error.issues
      else parsedData.body = parsed.data as z.infer<TBody>
    }

    if (errors.body || errors.query) {
      if (errorHandler) {
        errorHandler(errors, event)
        return
      }

      throw createError({
        statusCode: 400,
        statusMessage: JSON.stringify({ errors }),
      })
    }

    event.context.parsedData = parsedData

    return handler(event)
  })
}
