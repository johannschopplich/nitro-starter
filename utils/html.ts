/**
 * @remarks
 * Tip: Use the `lit-html` VSCode extension to get syntax highlighting
 */
export function html(...args: Parameters<typeof String.raw>) {
  return String.raw(...args)
}
