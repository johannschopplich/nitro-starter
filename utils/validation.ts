import { getValidatedQuery } from 'h3'
import { z } from 'zod'
import type { H3Event } from 'h3'

type UnknownKeysParam = 'passthrough' | 'strict' | 'strip'

type Schema<U extends UnknownKeysParam = any> =
  | z.ZodObject<any, U>
  | z.ZodUnion<[Schema<U>, ...Schema<U>[]]>
  | z.ZodIntersection<Schema<U>, Schema<U>>
  | z.ZodDiscriminatedUnion<string, z.ZodObject<any, U>[]>
  | z.ZodEffects<z.ZodTypeAny>

type ParsedData<T extends Schema | z.ZodRawShape> = T extends Schema
  ? z.output<T>
  : T extends z.ZodRawShape
  ? z.output<z.ZodObject<T>>
  : never

export function useValidatedQuery<T extends Schema | z.ZodRawShape>(
  event: H3Event,
  schema: T,
): Promise<ParsedData<T>> {
  return getValidatedQuery(event, validateSchemaFn(schema))
}

export function useValidatedBody<T extends Schema | z.ZodRawShape>(
  event: H3Event,
  schema: T,
): Promise<ParsedData<T>> {
  return readValidatedBody(event, validateSchemaFn(schema))
}

function validateSchemaFn<T extends Schema | z.ZodRawShape>(schema: T) {
  return async function (data: unknown) {
    const _schema = schema instanceof z.ZodType ? schema : z.object(schema)

    try {
      return await _schema.parseAsync(data)
    } catch (error) {
      // Stringify zod error to avoid circular structure
      throw new Error(JSON.stringify(error))
    }
  }
}
