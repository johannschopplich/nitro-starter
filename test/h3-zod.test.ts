import { describe, expect, it } from 'vitest'
import { callHandler } from './utils'

describe('validate requests using zod schemas', () => {
  describe('validate query', () => {
    it('returns 200 if query matches validation schema', async () => {
      const { data, status } = await callHandler('/api/validate/query?required')

      expect(status).toBe(200)
      expect(data).toMatchSnapshot()
    })

    it('throws 400 if query does not match validation schema', async () => {
      const { data, status } = await callHandler('/api/validate/query')

      expect(status).toBe(400)
      expect(data).toMatchSnapshot()
    })
  })

  describe('validate body', () => {
    it('returns 200 if body matches validation schema', async () => {
      const { data, status } = await callHandler('/api/validate/body', {
        method: 'POST',
        body: JSON.stringify({ required: true }),
      })

      expect(status).toBe(200)
      expect(data).toMatchSnapshot()
    })

    it('throws 400 if body does not match validation schema', async () => {
      const { data, status } = await callHandler('/api/validate/body', {
        method: 'POST',
        body: JSON.stringify({}),
      })

      expect(status).toBe(400)
      expect(data).toMatchSnapshot()
    })
  })

  describe('event handler with schema', () => {
    it('returns 200 if query/body matches validation schema', async () => {
      const { data, status } = await callHandler(
        '/api/validate/handler?required=true',
        {
          method: 'POST',
          body: JSON.stringify({ required: true }),
        },
      )

      expect(status).toBe(200)
      expect(data).toMatchSnapshot()
    })

    it('throws 400 if query/body does not match validation schema', async () => {
      const { data, status } = await callHandler(
        '/api/validate/handler?required=true',
        { method: 'POST' },
      )

      expect(status).toEqual(400)
      expect(data).toMatchSnapshot()
    })

    it('passes parsed data to "event.context"', async () => {
      const { data } = await callHandler(
        '/api/validate/handler?required=true',
        {
          method: 'POST',
          body: JSON.stringify({ required: true }),
        },
      )

      expect(data.query.required).toBe(true)
      expect(data.body.required).toBe(true)
    })
  })
})
