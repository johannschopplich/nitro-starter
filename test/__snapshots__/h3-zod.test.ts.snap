// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`validate requests using zod schemas > validate body > returns 200 if body matches validation schema 1`] = `
{
  "required": true,
}
`;

exports[`validate requests using zod schemas > validate body > throws 400 if body does not match validation schema 1`] = `
{
  "message": "Error",
  "statusCode": 400,
  "statusMessage": "Bad Request",
  "url": "/api/validate/body",
}
`;

exports[`validate requests using zod schemas > validate query > returns 200 if query matches validation schema 1`] = `
{
  "user": "nitro",
}
`;

exports[`validate requests using zod schemas > validate query > throws 400 if query does not match validation schema 1`] = `
{
  "message": "Error",
  "statusCode": 400,
  "statusMessage": "Bad Request",
  "url": "/api/validate/query",
}
`;
